Python/chapter02/2.5 - Sum Lists/miguel_2.5_soln.py

```diff
@@ -0,0 +1,315 @@
+            # must point to same node
+            self.tail = new_head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def pop_head(self) -> Node:
+        if self.head is None:
+            raise IndexError('no head to pop')
+        h = self.head
+        h.next = None
+        self.head = self.head.next
+        self.size -= 1
+        return h
+
+    def append(self, ll: 'LinkedList') -> None:
+        self.tail.next = ll.head
+        self.size += ll.size
+        ll.head = None
+        ll.size = 0
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def sum_lists_forward(ll1: LinkedList, ll2: LinkedList) -> LinkedList:
+    """
+    Will add ll1 and ll2 where each element in the linked lists
+    represent digits.
+    Precondition: ll1 and ll2 must be the same length.
+    Digits are in forward order, for example:
+    Input: (6 -> 1 -> 7) + (2 -> 9 -> 5). That is, 617 + 295
+    Output: 9 -> 1 -> 2. That is, 912
+    Runtime: O(n)
+    Space Complexity: O(n) - output_ll
+    :param ll1: first input linked list
+    :param ll2: second input linked list
+    :return: a linked list containing the result of the addition
+    """
+    assert(ll1.size == ll2.size)
+    n1 = ll1.head
+    n2 = ll2.head
+    output_ll = LinkedList()
+    prev_node = None
+    while n1 is not None:
+        first = n1.data
+        second = n2.data
+        result = first + second
+        if result >= 10:
+            carry = 1
+            result -= 10  # extract digit in one's place
+            if prev_node is None:
+                # for when the first addition yields a carry
+                # and the output linked list is empty
+                output_ll.append_to_head(carry)
+            else:
+                # add carry to previous digit
+                prev_node.data += carry
+        output_ll.append_to_tail(result)
+        n1 = n1.next
+        n2 = n2.next
+        prev_node = output_ll.tail
+
+    # if carry > 0:
```

Comment:
i see that there are tests for this case, but how are they passing? O_o

Reply:
is there a bug in the algorithm that would or should prevent the test cases for forward sum from passing?

Reply:
Nevermind, I was confused about the fact that the carry code was commented out, but it was buried inside the loop.

it would have probably been better to add a break or something to be able to have each chunk of code do a single logical operation: the loop should have only dealt with the core addition loop.